os/kernel/log_dump/log_dump.c

```diff
@@ -183,52 +267,148 @@ int log_dump_save(char ch)
 			curbytes = 1;
-			current_ptr = 0; /* reset the read pointer as head is modified */
+			current_ptr = 0;	/* reset the read pointer as head is modified */
 			node->arr[0] = ch;
-			sq_addlast((sq_entry_t *)node, &log_dump_chunks);
+			sq_addlast((sq_entry_t *) node, &log_dump_chunks);
+			if (set_comp_head) {	/* new node, so compressed block starts at 0 */
+				node->comp_head = 0;
+				set_comp_head = false;
+			} else {
+				node->comp_head = LOG_DUMP_COMPHEAD_OFFSET;
+				set_comp_head = false;
+			}
 			log_dump_mem_check(max_size);
 		}
 
-	} else { /* there is place in tail */
-
+	} else {					/* there is place in tail */
+		struct log_dump_chunk_s *log_dump_tail = (struct log_dump_chunk_s *)sq_tail(&log_dump_chunks);
 		log_dump_tail->arr[curbytes] = ch;
 		curbytes += 1;
+		if (set_comp_head) {	/* new node, so compressed block starts at 0 */
+			if (log_dump_tail->comp_head < 0) {	/* compressed node not present */
+				log_dump_tail->comp_head = curbytes - 1;
+			}
+			set_comp_head = false;
+		}
 		if (curbytes % CONFIG_LOG_DUMP_MEMCHECK_SIZE == 0) {
 			log_dump_mem_check(max_size);
+			return LOG_DUMP_OK_HEAPCHK;
 		}
 	}
 
 	return LOG_DUMP_OK;
 }
 
+int log_dump_save(char ch)
+{
+	if (is_started_to_save == false) {
+		return LOG_DUMP_OK;
+	}
+
+	size_t free_size = 0;
+	int ret = 0;
+	uncomp_array[uncomp_curbytes] = ch;
+	uncomp_curbytes++;
+	if (uncomp_curbytes == CONFIG_LOG_DUMP_CHUNK_SIZE) {
+		/* compress the block, add it to the nodes, reset the uncomp_array */
+		set_comp_head = true;
+
+		out_buf = compress_allocate_buffer(0, CONFIG_LOG_DUMP_CHUNK_SIZE);
+
+		if (out_buf == NULL) {
+			lldbg("memory allocation failure\n");
+			return LOG_DUMP_MEM_FAIL;
+		}
+
+		current_compression = true;
+
+		struct log_dump_msg_s msg;
+		msg.comp = false;
+		msg.ch = 'c';
+		mq_send(log_dump_get_mqfd(), (const char *)&msg, sizeof(struct log_dump_msg_s), 100);
+
+		while (current_compression) ;
+
+		if (compress_ret < 0) {
+			lldbg("Fail to compress ret = %d\n", compress_ret);
+			kmm_free(out_buf);
+			return compress_ret;
+		}
+		uncomp_curbytes = 0;	/* reset */
+		snprintf(comp_size, LOG_DUMP_COMPRESS_NODESZ, "%04d", writesize);
+		for (int i = 0; i < LOG_DUMP_COMPRESS_NODESZ - 1; i++) {
+			ret = log_dump_tobuffer(comp_size[i], &free_size);
+			if (ret < 0) {
+				kmm_free(out_buf);
+				return ret;
+			}
+		}
+
+		for (int i = 0; i < writesize; i++) {
+			ret = log_dump_tobuffer(out_buf[i], &free_size);
+			if (ret < 0) {
+				kmm_free(out_buf);
+				return ret;
+			}
+		}
+		kmm_free(out_buf);
+		ret = LOG_DUMP_OK_HEAPCHK;
+
+	} else {
+		ret = LOG_DUMP_OK;
+	}
+
+	if (ret == LOG_DUMP_OK_HEAPCHK) {
+		char fsz_array[LOG_DUMP_STRLEN_FREEHEAP];	/* For Free heap memory string */
+		snprintf(fsz_array, sizeof(size_t), "<fsz:%08d>", free_size);
+		for (int i = 0; i < sizeof(fsz_array); i++) {
+			uncomp_array[uncomp_curbytes++] = fsz_array[i];
+		}
+	}
```

Comment:
I think you can move this into the if above. Then no need to distinguish OK_HEAP I think. Could you check?

Reply:
done